### ISSUE TITLE
Merging `master` into `dev` clashes with protected `dev`

### ISSUE THREAD

**Issue by thijsdhollander** [MRtrix3/mrtrix3#1120]
Little intro:

Using the git flow model, every change that happens to `master`, should in principle immediately be reflected on `dev` too (in proper git flow lingo/theory, this would be the hot fix scenario (we've used it ourselves here and there lately for things that aren't really that "hot" to fix; but that's besides the point of what I'm raising in this issue).  We were lagging a bit behind, with the risk of `master` and `dev` diverging too much.  Immediately reflecting hot fixes or other changes to `master` on `dev` as well guarantees that feature branches being developed on top of `dev` also work with everything that's happened on `master`.  By the time a release comes around, the merge of the release branch (branched from `dev`) to `master` should then be trivial, since the release that sits ready to be merged with `master` is already up to date with `master`.  Ok, so far just making sure we're all talking about the same scenario.

The problem:

I was just trying to bring `dev` up to date with the most recent few changes on `master` (that happened due to features/fixes that were deemed appropriate to go straight to `master`; the hot fix scenario essentially).  As mentioned above, we should've in principle done that immediately for each of those hot fix type of things that went to `master`; but ok, I thought I'd just do it now before things diverge too much.  That's where the problem happened.  I first made sure I'd pulled the latest `dev` locally, and made sure the updated test data reference to the test data repo was there as well.  Then I merged `origin/master` locally in `dev`, giving this output:

```
Auto-merging lib/mrtrix3/app.py

Merge made by the 'recursive' strategy.
 Doxyfile                               |  9 +++++----
 core/file/config.cpp                   | 14 +++++++++-----
 core/image_diff.h                      |  6 +++---
 lib/mrtrix3/app.py                     |  2 +-
 src/gui/mrview/tool/roi_editor/roi.cpp | 60 ++++++++++++++++++++++++++++++------------------------------
 5 files changed, 48 insertions(+), 43 deletions(-)
```

...all as expected: these are only those few changes that happened to `master` lately indeed.  However, upon trying to push the merge commit, I was then greeted by:

```
Counting objects: 8, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (7/7), done.
Writing objects: 100% (8/8), 1.43 KiB | 0 bytes/s, done.
Total 8 (delta 6), reused 0 (delta 0)
remote: Resolving deltas: 100% (6/6), completed with 6 local objects.
remote: error: GH006: Protected branch update failed for refs/heads/dev.
remote: error: Required status check "continuous-integration/travis-ci" is expected.
To git@github.com:MRtrix3/mrtrix3.git
 ! [remote rejected] dev -> dev (protected branch hook declined)
error: failed to push some refs to 'git@github.com:MRtrix3/mrtrix3.git'
```

I do remember successfully updating `dev` recently without issues though, but that was still before many of the recent features had been merged to `dev`, so those may have not been merge commits.  The other difference (since then) I can think of is the changing of some test (data) related things on `dev`, where it does sort of make sense that the `dev` branch protection would do its job.

But well, so the question remains: is there a proper way to get the latest changes of `master` on `dev`, taking into account the particular protection setup we've got going?  We do need this to get something git flow-ish working... :confused:

**Comment by thijsdhollander** [comment]
Quick one before I leave work for the day: it's not particularly clean, but something that comes to mind would be branching (current) `master` and merging that `whatever_branch` into `dev` via a proper pull request, which will then also require to merge `dev` into `whatever_branch`, so the tests from `dev` can be run properly.

Or, now that I think of it, immediately merging each `hotfix_branch` into `dev` via a separate pull request after the pull request of `hotfix_branch` into `master` has been done, including merging `dev` into that `hotfix_branch` just after the merge to `master` and running the tests.  So rather than `hotfix_branch` to `master` followed by `master` to `dev`; instead `hotfix_branch` to `master` followed by `hotfix_branch` to `dev`.  But to do that properly, we do need to take responsibility to immediately do this for a hotfix, rather than after a series of them (as is now the case).

Does some of the above make sense, or am I missing something?

**Comment by jdtournier** [comment]
Yes, it's a little messier since I protected the `dev` branch too. But it's a simple matter of creating a fresh pull request for it - which I've just done in #1121. This is what I did (note I use `github` instead of `origin` for the main repo):

- checkout `dev` branch:
  ```ShellSession
  $ git checkout dev
  Switched to branch 'dev'
  Your branch is up-to-date with 'github/dev'.
  ```

- make sure everything is up to date:
  ```ShellSession
  $ git fetch --all
  ```

- create new branch from `dev`:
  ```ShellSession
  $ git checkout -b update_dev_from_master
  Switched to a new branch 'update_dev_from_master'
  ```

- do the merge:
  ```ShellSession
  $ git merge github/master 
  Auto-merging lib/mrtrix3/app.py
  Merge made by the 'recursive' strategy.
   Doxyfile                               |  9 +++++----
   core/file/config.cpp                   | 14 +++++++++-----
   core/image_diff.h                      |  6 +++---
   lib/mrtrix3/app.py                     |  2 +-
   src/gui/mrview/tool/roi_editor/roi.cpp | 60 ++++++++++++++++++++++++++++++------------------------------
   5 files changed, 48 insertions(+), 43 deletions(-)
  ```

- push to new branch on GitHub:
  ```ShellSession
  $ git push -u github update_dev_from_master 
  Counting objects: 8, done.
  Delta compression using up to 12 threads.
  Compressing objects: 100% (6/6), done.
  Writing objects: 100% (8/8), 800 bytes | 800.00 KiB/s, done.
  Total 8 (delta 6), reused 1 (delta 1)
  remote: Resolving deltas: 100% (6/6), completed with 6 local objects.
  To github.com:MRtrix3/mrtrix3.git
   * [new branch]        update_dev_from_master -> update_dev_from_master
  Branch update_dev_from_master set up to track remote branch update_dev_from_master from github.
  ```

-  then create the pull request from the GitHub UI. 
 
A bit messy, I agree. But then it's not a bad thing to allow TravisCI to check that the merge hasn't screwed anything up on `dev` either...

**Comment by jdtournier** [comment]
In terms of managing this day to day, if we do want to have each hotfix on `master` merged into `dev` at around the same time: I'm not sure there's any way around having two pull requests, one for each branch... But it would be sufficient to merge `master` back into `dev` once the hotfix has been merged into it, which would ensure the two branches remain in sync, ready for a future merge. I don't see a particularly simpler way of handling this, maybe others have ideas on that front...?

**Comment by thijsdhollander** [comment]
Ok, all that makes full sense indeed.  It's quite clear that an extra pull request can't be avoided, since tests will only run and results reported via a pull request.  And since tests, and the code base in which they operate, can have changed on `dev` versus `master`, we're indeed talking about 2 entirely separate test runs.

In practice though, to avoid creating another branch (as was at this point in time the most convenient, but messy, solution to deal with a "backlog of hot fixes") I suppose the cleanest way is along the lines of what I mentioned somewhere above.  So the entire scenario for a hot fix from start to end would be (and closely follows the git flow branching and merging diagram for a hot fix):

1. branch from `master` to obtain your fresh `my_hotfix` branch
2. implement the hot fix on that branch
3. create pull request with the intention to merge `my_hotfix` into `master`
4. see "master" tests run and get feedback from other developers, all via the pull request
5. if tests and feedback are positive, hit the button and merge into `master`
6. merge `dev` into `my_hotfix` branch
7. create pull request with the intention to merge `my_hotfix` into `dev`
8. see "dev" tests run and in rare cases get feedback from other developers (in case hot fix conflicts/interacts with some new feature/development that was not on `master` yet)
9. if tests and rare case feedback are positive, hit the button and merge into `dev`

Steps 6 to 9 should be fast and without notable issues in most cases.  If a hot fix to `master` is anticipated to be challenging to play ball with `dev`, this should optimally be detected (by us) already at step 4.  We don't want a hot fix scenario to be dangling around "stuck" at steps 6 to 9, because that risks "breaking" the git flow model, which would worst case lead to anarchy and cannibalism if we would be stuck on an island around that time. :scream: :grin: 

But essentially, in git flow, both the "hot fix" scenario, as well as the "release" scenario both require some kind of "joint" merging of either the hot fix or the release branch to `master` and `dev` "at the same time".  In case of the "release" scenario, this should never be a problem if all the other rules have been followed closely (`dev` always up to date and `release_branch`'s lifetime as short as possible (only drafting a release) and always up to date).  In case of the "hot fix" scenario, in the rare case described above, anticipation at step 4 is probably the key.

**Comment by jdtournier** [comment]
Yep, sounds about right. But I think your concerns around anarchy and cannibalism are probably unfounded... `master` is protected by both the TravisCI tests _and_ a review, but `dev` only requires the tests to pass. So the only thing that would hold up the merge to `dev` (your steps 6-9) would be the legitimate problem that the hotfix breaks something in `dev` - but you wouldn't be waiting on someone else to review your changes...

**Comment by thijsdhollander** [comment]
> So the only thing that would hold up the merge to dev (your steps 6-9) would be the legitimate problem that the hotfix breaks something in dev - but you wouldn't be waiting on someone else to review your changes...

Sure, I wasn't concerned in this case that others would be the cause of holding up such a merge, it was really only the scenario where a hot fix would have technical issues with something in `dev`, where the resolution may require some help with another developer (e.g. the one who developed the feature on `dev` that doesn't play ball with the hot fix).  So both developers of hot fixes as well as developers of features (on `dev`) should probably try and keep an eye on hot fixes that are due to be finalised, and anticipate issues of the hot fix versus `dev` in the rare cases where such issues could exist, and optimally flag (and resolve) that at step 4.  But again, I can't imagine this happening frequently, so we should be good. :+1: 

So at the end of the day, the most important thing for people to do ("responsibly") is to make sure that they take their hot fixes all the way through these steps.

**Comment by jdtournier** [comment]
We'll need to add some info about all this somewhere obvious soon... I'm thinking of adding a `contributing.md` file alongside the `README.md`, what do you reckon? It would describe the git flow for various things like this, the branch protection, and the fork & pull request model for external contributors. Probably also include a link to the online dev doc. And we could refer to this document in the user docs too.

**Comment by thijsdhollander** [comment]
That's a good idea indeed.  It'd be best to move the information on the tests (e.g. how to update test data) into such a `contributing.md` as well; we've struggled with that scenario at times in the past as well... I don't see the test (data) being modified in a hot fix realistically, but it could (and will) of course for a feature branch.  So we need to check how such changes happen all the way from a feature branch, via `dev`, via a release branch, finally to `master`... taking into account things can sit for a while in the feature branch and `dev` branch stages (but move relatively quickly through the release branch stage).
